File: arrays_hashing/product_of_array_except_self/product_array_except_self.go
// Package productofarrayexceptself provides a function to calculate the product of an array excluding each element.
package productofarrayexceptself

func ProductExceptSelf(nums []int) []int {
	result := make([]int, len(nums))
	for idx := range result {
		result[idx] = 1
	}

	prefix := 1
	for i := range nums {
		result[i] = prefix
		prefix *= nums[i]
	}

	postfix := 1
	for i := range nums {
		end := len(nums) - 1 - i
		result[end] *= postfix
		postfix *= nums[end]
	}
	return result
}

File: arrays_hashing/product_of_array_except_self/product_array_except_self.py
def product_except_self(nums):
    result = [1] * len(nums)
    prefix = 1
    for i in range(len(nums)):
        result[i] = prefix
        prefix *= nums[i]
    postfix = 1
    for i in range(len(nums), 0, -1):
        result[i - 1] *= postfix
        postfix *= nums[i - 1]
    return result
